File: frontend/lib/api/broker.ts
export type BrokerSession = {
  id: string
  title: string
  status: string
  summary: string | null
  created_at: string
  updated_at: string
}

export type BrokerAttachment = {
  id: string
  session_id: string
  message_id?: string | null
  kind: "file" | "url"
  share_class: "pending" | "public" | "personal"
  label: string | null
  purpose: string | null
  original_filename: string | null
  content_type: string | null
  size_bytes: number | null
  url: string | null
  content_url: string | null
  description: string | null
  status: string
  created_at: string
}

export type BrokerUploadRequest = {
  id: string
  purpose: string
  suggested_share_class: string
  hint: string
  status: string
  created_at: string
}

export type BrokerMessage = {
  id: string
  session_id: string
  role: "assistant" | "user"
  content: string
  created_at: string
  attachments?: BrokerAttachment[]
}

export type BrokerPartyConnection = {
  id: string
  match_id: string
  source_user_id: string
  candidate_user_id: string
  status: string
  created_at: string
  updated_at: string
}

export type ConnectionPeer = {
  user_id: string
  name: string
  location: string
  request_title: string | null
  request_summary: string | null
}

export type ConnectionAttachment = {
  id: string
  connection_id: string
  message_id: string | null
  kind: "file" | "image" | string
  original_filename: string | null
  content_type: string | null
  size_bytes: number | null
  content_url: string | null
  created_at: string
}

export type ConnectionMessage = {
  id: string
  connection_id: string
  sender_user_id: string | null
  kind: "user" | "system" | string
  mine: boolean
  content: string
  created_at: string
  attachments: ConnectionAttachment[]
}

export type ConnectionSummary = {
  id: string
  match_id: string
  status: string
  peer: ConnectionPeer
  last_message: ConnectionMessage | null
  unread_count: number
  created_at: string
  updated_at: string
}

export type ConnectionDetail = {
  connection: ConnectionSummary
  messages: ConnectionMessage[]
}

export type BrokerSessionDetail = {
  session: BrokerSession
  messages: BrokerMessage[]
  attachments?: BrokerAttachment[]
  pending_upload_requests?: BrokerUploadRequest[]
}

export function getBrokerBackendUrl() {
  return (
    process.env.NEXT_PUBLIC_BACKEND_URL ||
    process.env.BACKEND_URL ||
    "http://127.0.0.1:8000"
  )
}

function getBackendUrl() {
  return getBrokerBackendUrl()
}

export function brokerWebsocketUrl(accessToken: string) {
  const http = getBrokerBackendUrl().replace(/\/$/, "")
  const ws = http.replace(/^https:/i, "wss:").replace(/^http:/i, "ws:")
  return `${ws}/broker/ws?access_token=${encodeURIComponent(accessToken)}`
}

async function brokerFetch<T>(
  accessToken: string,
  path: string,
  init?: RequestInit
) {
  const response = await fetch(`${getBackendUrl()}${path}`, {
    ...init,
    headers: {
      Authorization: `Bearer ${accessToken}`,
      "Content-Type": "application/json",
      ...init?.headers
    }
  })

  if (!response.ok) {
    const body = await response.json().catch(() => null)
    const detail =
      typeof body?.detail === "string"
        ? body.detail
        : `Broker API failed with status ${response.status}`
    throw new Error(detail)
  }

  if (response.status === 204) {
    return undefined as T
  }

  return (await response.json()) as T
}

export function listBrokerSessions(accessToken: string) {
  return brokerFetch<BrokerSession[]>(accessToken, "/broker/sessions")
}

export function createBrokerSession(
  accessToken: string,
  initialMessage?: string
) {
  return brokerFetch<BrokerSessionDetail>(accessToken, "/broker/sessions", {
    method: "POST",
    body: JSON.stringify({ initial_message: initialMessage || null })
  })
}

export function getBrokerSession(accessToken: string, sessionId: string) {
  return brokerFetch<BrokerSessionDetail>(
    accessToken,
    `/broker/sessions/${sessionId}`
  )
}

export function deleteBrokerSession(accessToken: string, sessionId: string) {
  return brokerFetch<void>(accessToken, `/broker/sessions/${sessionId}`, {
    method: "DELETE"
  })
}

export function sendBrokerMessage(
  accessToken: string,
  sessionId: string,
  content: string,
  attachmentIds: string[] = []
) {
  return brokerFetch<BrokerMessage[]>(
    accessToken,
    `/broker/sessions/${sessionId}/messages`,
    {
      method: "POST",
      body: JSON.stringify({
        content,
        attachment_ids: attachmentIds
      })
    }
  )
}

export async function uploadBrokerAttachment(
  accessToken: string,
  sessionId: string,
  file: File,
  options?: { caption?: string; requestId?: string }
) {
  const body = new FormData()
  body.append("file", file)
  if (options?.caption) body.append("caption", options.caption)
  if (options?.requestId) body.append("request_id", options.requestId)

  const response = await fetch(
    `${getBackendUrl()}/broker/sessions/${sessionId}/attachments`,
    {
      method: "POST",
      headers: {
        Authorization: `Bearer ${accessToken}`
      },
      body
    }
  )
  if (!response.ok) {
    const payload = await response.json().catch(() => null)
    const detail =
      typeof payload?.detail === "string"
        ? payload.detail
        : `Upload failed with status ${response.status}`
    throw new Error(detail)
  }
  return (await response.json()) as BrokerAttachment
}

export function addBrokerLink(
  accessToken: string,
  sessionId: string,
  url: string,
  options?: { caption?: string; requestId?: string }
) {
  return brokerFetch<BrokerAttachment>(
    accessToken,
    `/broker/sessions/${sessionId}/links`,
    {
      method: "POST",
      body: JSON.stringify({
        url,
        caption: options?.caption || null,
        request_id: options?.requestId || null
      })
    }
  )
}

export function grantBrokerAttachment(
  accessToken: string,
  attachmentId: string,
  matchId: string,
  granted: boolean
) {
  return brokerFetch<BrokerAttachment>(
    accessToken,
    `/broker/attachments/${attachmentId}/grants`,
    {
      method: "POST",
      body: JSON.stringify({ match_id: matchId, granted })
    }
  )
}

export function connectBrokerMatch(accessToken: string, matchId: string) {
  return brokerFetch<ConnectionSummary>(
    accessToken,
    `/broker/matches/${matchId}/connect`,
    {
      method: "POST"
    }
  )
}

export function listBrokerConnections(accessToken: string) {
  return brokerFetch<ConnectionSummary[]>(accessToken, "/broker/connections")
}

export function getBrokerConnection(accessToken: string, connectionId: string) {
  return brokerFetch<ConnectionDetail>(
    accessToken,
    `/broker/connections/${connectionId}`
  )
}

export function sendConnectionMessage(
  accessToken: string,
  connectionId: string,
  content: string,
  attachmentIds: string[] = []
) {
  return brokerFetch<ConnectionMessage>(
    accessToken,
    `/broker/connections/${connectionId}/messages`,
    {
      method: "POST",
      body: JSON.stringify({
        content,
        attachment_ids: attachmentIds
      })
    }
  )
}

export async function uploadConnectionAttachment(
  accessToken: string,
  connectionId: string,
  file: File
) {
  const body = new FormData()
  body.append("file", file)

  const response = await fetch(
    `${getBackendUrl()}/broker/connections/${connectionId}/attachments`,
    {
      method: "POST",
      headers: {
        Authorization: `Bearer ${accessToken}`
      },
      body
    }
  )
  if (!response.ok) {
    const payload = await response.json().catch(() => null)
    const detail =
      typeof payload?.detail === "string"
        ? payload.detail
        : `Upload failed with status ${response.status}`
    throw new Error(detail)
  }
  return (await response.json()) as ConnectionAttachment
}

export function markConnectionRead(accessToken: string, connectionId: string) {
  return brokerFetch<ConnectionSummary>(
    accessToken,
    `/broker/connections/${connectionId}/read`,
    { method: "POST" }
  )
}

export function getPushPublicKey(accessToken: string) {
  return brokerFetch<{ public_key: string; configured: boolean }>(
    accessToken,
    "/broker/push/vapid-public-key"
  )
}

export function subscribePush(
  accessToken: string,
  subscription: {
    endpoint: string
    keys: { p256dh: string; auth: string }
    user_agent?: string
  }
) {
  return brokerFetch<void>(accessToken, "/broker/push/subscribe", {
    method: "POST",
    body: JSON.stringify(subscription)
  })
}

File: backend-agent/app/schemas/broker.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, model_validator


class BrokerSessionCreate(BaseModel):
    initial_message: str | None = Field(default=None, max_length=4000)


class BrokerMessageCreate(BaseModel):
    content: str = Field(default="", max_length=4000)
    attachment_ids: list[str] = Field(default_factory=list)

    @model_validator(mode="after")
    def require_text_or_attachments(self) -> BrokerMessageCreate:
        if not self.content.strip() and not self.attachment_ids:
            raise ValueError("Send a message or at least one attachment.")
        return self


class BrokerLinkCreate(BaseModel):
    url: str = Field(min_length=8, max_length=2000)
    caption: str | None = Field(default=None, max_length=400)
    request_id: str | None = None


class BrokerAttachmentPatch(BaseModel):
    share_class: Literal["public", "personal"] | None = None
    label: str | None = Field(default=None, max_length=160)
    purpose: str | None = Field(default=None, max_length=32)


class BrokerAttachmentGrantCreate(BaseModel):
    match_id: str
    granted: bool = True


class BrokerAttachmentRead(BaseModel):
    id: str
    session_id: str
    kind: str
    share_class: str
    label: str | None
    purpose: str | None
    original_filename: str | None
    content_type: str | None
    size_bytes: int | None
    url: str | None = None
    content_url: str | None = None
    description: str | None
    status: str
    created_at: datetime

    model_config = {"from_attributes": True}

    @classmethod
    def from_attachment(
        cls,
        attachment: Any,
        *,
        include_url: bool = False,
        content_url: str | None = None,
    ) -> BrokerAttachmentRead:
        return cls(
            id=attachment.id,
            session_id=attachment.session_id,
            kind=attachment.kind,
            share_class=attachment.share_class,
            label=attachment.label,
            purpose=attachment.purpose,
            original_filename=attachment.original_filename,
            content_type=attachment.content_type,
            size_bytes=attachment.size_bytes,
            url=attachment.url if include_url else None,
            content_url=content_url,
            description=attachment.description,
            status=attachment.status,
            created_at=attachment.created_at,
        )


class BrokerAttachmentRequestRead(BaseModel):
    id: str
    purpose: str
    suggested_share_class: str
    hint: str
    status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class BrokerSessionRead(BaseModel):
    id: str
    title: str
    status: str
    summary: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class BrokerMessageRead(BaseModel):
    id: str
    session_id: str
    role: str
    content: str
    created_at: datetime
    attachments: list[BrokerAttachmentRead] = Field(default_factory=list)

    model_config = {"from_attributes": True}

    @classmethod
    def from_message(
        cls,
        message: Any,
        attachments: list[BrokerAttachmentRead] | None = None,
    ) -> BrokerMessageRead:
        return cls(
            id=message.id,
            session_id=message.session_id,
            role=message.role,
            content=message.content,
            created_at=message.created_at,
            attachments=attachments or [],
        )


class BrokerRequestRead(BaseModel):
    id: str
    session_id: str
    user_id: str
    status: str
    title: str
    summary: str
    details: dict
    indexed: bool
    waiting_match_id: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

    @classmethod
    def from_request(cls, request) -> BrokerRequestRead:
        return cls(
            id=request.id,
            session_id=request.session_id,
            user_id=request.user_id,
            status=request.status,
            title=request.title,
            summary=request.summary,
            details=_public_details(request.details or {}),
            indexed=request.indexed_at is not None,
            waiting_match_id=request.waiting_match_id,
            created_at=request.created_at,
            updated_at=request.updated_at,
        )


class BrokerSessionDetail(BaseModel):
    session: BrokerSessionRead
    messages: list[BrokerMessageRead]
    request: BrokerRequestRead | None = None
    attachments: list[BrokerAttachmentRead] = Field(default_factory=list)
    pending_upload_requests: list[BrokerAttachmentRequestRead] = Field(default_factory=list)


class BrokerRequestMatch(BaseModel):
    request_id: str
    title: str
    summary: str
    status: str
    distance: float | None


class BrokerMatchRead(BaseModel):
    id: str
    source_request_id: str
    candidate_request_id: str
    status: str
    close_reason: str | None
    score: float | None
    match_reason: str | None
    last_activity_at: datetime
    expires_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class BrokerEventRead(BaseModel):
    id: str
    match_id: str
    session_id: str | None
    user_id: str | None
    event_type: str
    message_text: str | None
    payload: dict | None
    created_at: datetime

    model_config = {"from_attributes": True}


class BrokerPartyConnectionRead(BaseModel):
    id: str
    match_id: str
    source_user_id: str
    candidate_user_id: str
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ConnectionPeerRead(BaseModel):
    user_id: str
    name: str
    location: str
    request_title: str | None = None
    request_summary: str | None = None


class ConnectionAttachmentRead(BaseModel):
    id: str
    connection_id: str
    message_id: str | None
    kind: str
    original_filename: str | None
    content_type: str | None
    size_bytes: int | None
    content_url: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class ConnectionMessageRead(BaseModel):
    id: str
    connection_id: str
    sender_user_id: str | None
    kind: str
    mine: bool = False
    content: str
    created_at: datetime
    attachments: list[ConnectionAttachmentRead] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class ConnectionSummaryRead(BaseModel):
    id: str
    match_id: str
    status: str
    peer: ConnectionPeerRead
    last_message: ConnectionMessageRead | None = None
    unread_count: int = 0
    created_at: datetime
    updated_at: datetime


class ConnectionDetailRead(BaseModel):
    connection: ConnectionSummaryRead
    messages: list[ConnectionMessageRead]


class ConnectionMessageCreate(BaseModel):
    content: str = Field(default="", max_length=4000)
    attachment_ids: list[str] = Field(default_factory=list)

    @model_validator(mode="after")
    def require_text_or_attachments(self) -> ConnectionMessageCreate:
        if not self.content.strip() and not self.attachment_ids:
            raise ValueError("Send a message or at least one attachment.")
        return self


class PushSubscriptionCreate(BaseModel):
    endpoint: str = Field(min_length=8, max_length=2000)
    keys: dict[str, str]
    user_agent: str | None = Field(default=None, max_length=255)

    @model_validator(mode="after")
    def require_push_keys(self) -> PushSubscriptionCreate:
        p256dh = (self.keys or {}).get("p256dh")
        auth = (self.keys or {}).get("auth")
        if not p256dh or not auth:
            raise ValueError("Push subscription keys must include p256dh and auth.")
        return self


class PushUnsubscribe(BaseModel):
    endpoint: str = Field(min_length=8, max_length=2000)


class VapidPublicKeyRead(BaseModel):
    public_key: str
    configured: bool = True


def _public_details(details: dict) -> dict:
    return {key: value for key, value in details.items() if key != "semantic"}
